src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,414 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::errors::CredentialError;
+use crate::token::{Token, TokenProvider};
+use crate::Result;
+use std::sync::{Arc, RwLock};
+use tokio::sync::{Mutex, Notify};
+// Using tokio's wrapper makes the cache testable without relying on clock times.
+use tokio::time::Instant;
+
+// TODO(#1210) - implementation in progress
+#[allow(dead_code)]
+#[derive(Debug)]
+pub(crate) struct TokenCache<T>
+where
+    T: TokenProvider,
+{
+    // The cached token, or the last seen error.
+    token: Arc<RwLock<Result<Token>>>,
+
+    // Tracks if a refresh is ongoing. If the lock is held, there is a refresh.
+    refresh_in_progress: Arc<Mutex<()>>,
+    // Allows us to await the result of a refresh in multiple threads.
+    refresh_notify: Arc<Notify>,
+
+    // The token provider. This thing does the refreshing.
+    inner: Arc<T>,
+}
+
+// We manually implement the `Clone` trait because the Rust compiler will
+// squawk if `T` is not `Clone`, even though we only hold an `Arc<T>`. :shrug:
+impl<T: TokenProvider> Clone for TokenCache<T> {
+    fn clone(&self) -> TokenCache<T> {
+        TokenCache {
+            token: self.token.clone(),
+            refresh_in_progress: self.refresh_in_progress.clone(),
+            refresh_notify: self.refresh_notify.clone(),
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+fn initial_token() -> Result<Token> {
+    Err(CredentialError::retryable_from_str("No token or error in the cache. This should never happen. Something has gone wrong. Open an issue at https://github.com/googleapis/google-cloud-rust/issues/new?template=bug_report.md."))
+}
+
+impl<T: TokenProvider> TokenCache<T> {
+    // TODO(#1210) - implementation in progress
+    #[allow(dead_code)]
+    pub fn new(inner: T) -> TokenCache<T> {
+        TokenCache {
+            token: Arc::new(RwLock::new(initial_token())),
+            refresh_in_progress: Arc::new(Mutex::new(())),
+            refresh_notify: Arc::new(Notify::new()),
```

Comment:
I think this is getting sufficiently complicated that it needs a design doc, maybe one or two pages or a detailed GitHub issue would do.

Reply:
What do you want clarified in the design doc? Is it more the specification or the implementation?

Reply:
The specification would help.  I think a description of why you are using whatever synchronization primitives you are using would help too.

- What alternatives exist?
- Why did we pick the ones we did?